modules/koblerBidAdapter.js

```diff
@@ -0,0 +1,227 @@
+import {BANNER} from '../src/mediaTypes.js';
+
+const BIDDER_CODE = 'kobler';
+const BIDDER_ENDPOINT = 'https://bid.essrtb.com/bid/prebid_rtb_call';
+const TIMEOUT_NOTIFICATION_ENDPOINT = 'https://bid.essrtb.com/notify/prebid_timeout';
+const DEFAULT_CURRENCY = 'USD';
+const DEFAULT_TIMEOUT = 1000;
+const TIME_TO_LIVE_IN_SECONDS = 10 * 60;
+
+export const isBidRequestValid = function (bid) {
+  return !(!bid || !bid.bidId || !bid.params || !bid.params.placementId);
+};
+
+export const buildRequests = function (validBidRequests, bidderRequest) {
+  return {
+    method: 'POST',
+    url: BIDDER_ENDPOINT,
+    data: buildOpenRtbBidRequestPayload(validBidRequests, bidderRequest),
+    options: {
+      contentType: 'application/json'
+    }
+  };
+};
+
+export const interpretResponse = function (serverResponse) {
+  const res = serverResponse.body;
+  const bids = []
+  if (res) {
+    res.seatbid.forEach(sb => {
+      sb.bid.forEach(b => {
+        bids.push({
+          requestId: b.impid,
+          cpm: b.price,
+          currency: res.cur,
+          width: b.w,
+          height: b.h,
+          creativeId: b.crid,
+          dealId: b.dealid,
+          netRevenue: true,
+          ttl: TIME_TO_LIVE_IN_SECONDS,
+          ad: b.adm,
+          nurl: b.nurl,
+          meta: {
+            advertiserDomains: b.adomain
+          }
+        })
+      })
+    });
+  }
+  return bids;
+};
+
+export const onBidWon = function (bid) {
+  const cpm = bid.cpm || 0;
+  const adServerPrice = utils.deepAccess(bid, 'adserverTargeting.hb_pb') || 0;
+  if (utils.isStr(bid.nurl) && bid.nurl !== '') {
+    const winNotificationUrl = utils.replaceAuctionPrice(bid.nurl, cpm)
+      .replace(/\${AD_SERVER_PRICE}/g, adServerPrice);
+    utils.triggerPixel(winNotificationUrl);
+  }
+};
+
+export const onTimeout = function (timeoutDataArray) {
+  if (utils.isArray(timeoutDataArray)) {
+    timeoutDataArray.forEach(timeoutData => {
+      const query = utils.parseQueryStringParameters({
+        ad_unit_code: timeoutData.adUnitCode,
+        auction_id: timeoutData.auctionId,
+        bid_id: timeoutData.bidId,
+        timeout: timeoutData.timeout,
+        placement_id: utils.deepAccess(timeoutData, 'params.0.placementId'),
+        page_url: window.location.href,
+      });
+      const timeoutNotificationUrl = `${TIMEOUT_NOTIFICATION_ENDPOINT}?${query}`;
+      utils.triggerPixel(timeoutNotificationUrl);
+    });
+  }
+};
+
+function buildOpenRtbBidRequestPayload(validBidRequests, bidderRequest) {
+  const imps = validBidRequests.map(br => buildOpenRtbImpObject(br));
+  const timeout = bidderRequest.timeout || config.getConfig('bidderTimeout') || DEFAULT_TIMEOUT;
+  const pageUrl = (bidderRequest.refererInfo && bidderRequest.refererInfo.referer)
+    ? bidderRequest.refererInfo.referer
+    : window.location.href;
+
+  const request = {
+    id: bidderRequest.auctionId,
+    at: 1, // TODO: is the auction always first price?
```

Comment:
As far as I know, Prebid.js is always first price. Is this correct? I am asking because some adapter set `at` to `2` which means the auction is second price.

Reply:
Most bidders set this to 1 if the result of the auction within their exchange is first price, notwithstanding the prebid or ad server auction.

Reply:
I think `at` should be used to indicate the type of the auction in Prebid since this bid request is sent from Prebid to the bidder. Perhaps the bidders that set it to 2 do so because the auction within their exchange is second price?

Reply:
That is my understanding as well on the behavior of other bidders.  

Prebid doesn't have built in functionality for a second price auction.

Reply:
Okay, thanks for clearing this up.